### PR TITLE
Generate Stack Traces for bad http responses

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

When we spit out a http reponse with a `0` http status code we
should log a trace back so we can easily find where things went wrong.

Change-Id: Ibb604d724fc8b660a2f16c7d14e5c0ad4932e4d0

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go]: avoid exporting this if you can

**Comment by liggitt** [comment]
given that there were no panics logged when we saw this response, the stack trace here is not likely to give us any stack info down into the nested handler that was actually responsible for returning early without writing a status code or content type

**Comment by dims** [comment]
the `Log()` in the `defer httplog.NewLogged(req, &w).StacktraceWhen(badHttpResponse()).Log()` is called when the actual serialization to stream takes place. so if at that point the http status is `0` or the `content-type` is not set then we will see a trace back.

worst case, we will not hit anything and we can revert this diagnostic change back...

**Comment by liggitt** [comment]
let's see if we can get any info in this PR before merging... would prefer not to change the signature if possible

**Comment by dims** [comment]
@sttts @liggitt given the low frequency of hits ... we won't catch the problem using `/test all` here. Is there something else we can do?

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=0-length%20response

**Comment by wojtek-t** [comment]
> given that there were no panics logged when we saw this response, the stack trace here is not likely to give us any stack info down into the nested handler that was actually responsible for returning early without writing a status code or content type

+100 - my thoughts exactly
I'm not sure this will provide us more details.

**Comment by dims** [comment]
@wojtek-t @liggitt fine. i'll just continue to investigate with stuff here

/hold

**Review by logicalhan — APPROVED** [review]
/lgtm

**Comment by wojtek-t** [comment]
LGTM - thanks @dims 
Let's hope we will get some results from it soon.

**Comment by liggitt** [comment]
no... https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/70439/pull-kubernetes-integration/40019/ failed:

```
--- FAIL: Test202StatusCode (3.75s)
    apiserver_test.go:140: Failed to create rs: 0-length response with status code: 200 and content type: 
```

and the "Bad HTTP Response" message did not get logged

**Pull request merged**
